sklearn/ensemble/_gb.py

```diff
@@ -28,19 +28,231 @@
+    search step in "Greedy Function Approximation" by Friedman.
+
+    Update equals:
+        loss.fit_intercept_only(y_true - raw_prediction)
+
+    Note: It only works, if the loss is a function of the residual, as is the
+    case for AbsoluteError and PinballLoss. Otherwise, one would need to get
+    the minimum of loss(y_true, raw_prediction + x) in x. A few examples:
+        - AbsoluteError: median(y_true - raw_prediction).
+        - PinballLoss: quantile(y_true - raw_prediction).
+    See also notes about need_update_leaves_values in BaseLoss.
+
+    Parameters
+    ----------
+    loss : BaseLoss
+    tree : tree.Tree
+        The tree object.
+    X : ndarray of shape (n_samples, n_features)
+        The data array.
+    y : ndarray of shape (n_samples,)
+        The target labels.
+    residual : ndarray of shape (n_samples,)
+        The residuals (usually the negative gradient).
+    raw_prediction : ndarray of shape (n_samples, n_trees_per_iteration)
+        The raw predictions (i.e. values from the tree leaves) of the
+        tree ensemble at iteration ``i - 1``.
+    sample_weight : ndarray of shape (n_samples,)
+        The weight of each sample.
+    sample_mask : ndarray of shape (n_samples,)
+        The sample mask to be used.
+    learning_rate : float, default=0.1
+        Learning rate shrinks the contribution of each tree by
+         ``learning_rate``.
+    k : int, default=0
+        The index of the estimator being updated.
+    """
+    # compute leaf for each sample in ``X``.
+    terminal_regions = tree.apply(X)
+
+    if not isinstance(loss, HalfSquaredError):
+        # mask all which are not in sample mask.
+        masked_terminal_regions = terminal_regions.copy()
+        masked_terminal_regions[~sample_mask] = -1
+
+        # update each leaf (= perform line search)
+        for leaf in np.nonzero(tree.children_left == TREE_LEAF)[0]:
+            indices = np.nonzero(terminal_regions == leaf)[0]  # of terminal regions
+            y_ = y.take(indices, axis=0)
+            sw = None if sample_weight is None else sample_weight[indices]
+
+            if isinstance(loss, HalfBinomialLoss):
+                # Make a single Newton-Raphson step.
+                # Our node estimate is given by:
+                #    sum(w * (y - prob)) / sum(w * prob * (1 - prob))
+                # we take advantage that: y - prob = residual
+                residual_ = residual.take(indices, axis=0)
+                prob = y_ - residual_
+                # numerator = negative gradient = y - prob
+                numerator = np.average(residual_, weights=sw)
+                # denominator = hessian = prob * (1 - prob)
+                denominator = np.average(prob * (1 - prob), weights=sw)
+
+                # prevents overflow and division by zero
+                sw_sum = indices.shape[0] if sw is None else np.sum(sw)
+                if abs(denominator * sw_sum) < 1e-150:
```

Comment:
I know this is trying to reproduce the original code by multipling by `np.sum(sw)`, but I think we can get away with just catching the warning:

```python
def _safe_divide(numerator, denominator):
    with np.errstate(divide="raise"):
        try:
            return numerator / denominator
        except FloatingPointError as e:
            return 0.0
```

and then using `_safe_divide` everywhere.

(Given that codecov is complaining, we do not even test for this edge case.)

Reply:
Good idea, done in 2f0a2930ca291ddb961464800c372cf9acc804db.